src/Components/Server/src/Circuits/DefaultPersistedCircuitCache.cs

```diff
@@ -0,0 +1,169 @@
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits;
+
+// Default implmentation of ICircuitPersistenceProvider that uses an in-memory cache
+internal sealed partial class DefaultInMemoryCircuitPersistenceProvider : ICircuitPersistenceProvider
+{
+    private readonly Lock _lock = new();
+    private readonly CircuitOptions _options;
+    private readonly MemoryCache _persistedCircuits;
+    private readonly Task<PersistedCircuitState> _noMatch = Task.FromResult<PersistedCircuitState>(null);
+    private readonly ILogger<ICircuitPersistenceProvider> _logger;
+
+    public PostEvictionCallbackRegistration PostEvictionCallback { get; internal set; }
+
+    public DefaultInMemoryCircuitPersistenceProvider(
+        ISystemClock clock,
+        ILogger<ICircuitPersistenceProvider> logger,
+        IOptions<CircuitOptions> options)
+    {
+        _options = options.Value;
+        _persistedCircuits = new MemoryCache(new MemoryCacheOptions
+        {
+            SizeLimit = _options.PersistedCircuitInMemoryMaxRetained,
+            Clock = clock
+        });
+
+        PostEvictionCallback = new PostEvictionCallbackRegistration
+        {
+            EvictionCallback = OnEntryEvicted
+        };
+
+        _logger = logger;
+    }
+
+    public Task PersistCircuitAsync(CircuitId circuitId, PersistedCircuitState persistedCircuitState, CancellationToken cancellation = default)
+    {
+        Log.CircuitPauseStarted(_logger, circuitId);
+
+        lock (_lock)
+        {
+            PersistCore(circuitId, persistedCircuitState);
+        }
+
+        return Task.CompletedTask;
+    }
+
+    private void PersistCore(CircuitId circuitId, PersistedCircuitState persistedCircuitState)
+    {
+        var cancellationTokenSource = new CancellationTokenSource(_options.PersistedCircuitInMemoryRetentionPeriod);
+        var options = new MemoryCacheEntryOptions
+        {
+            Size = 1,
+            PostEvictionCallbacks = { PostEvictionCallback },
+            ExpirationTokens = { new CancellationChangeToken(cancellationTokenSource.Token) },
+        };
+
+        var persistedCircuitEntry = new PersistedCircuitEntry
+        {
+            State = persistedCircuitState,
+            TokenSource = cancellationTokenSource
+        };
+
+        _persistedCircuits.Set(circuitId.Secret, persistedCircuitEntry, options);
+    }
+
+    private void OnEntryEvicted(object key, object value, EvictionReason reason, object state)
+    {
+        switch (reason)
+        {
+            case EvictionReason.Expired:
+            case EvictionReason.TokenExpired:
+            // Happens after the circuit state times out, this is triggered by the CancellationTokenSource we register
+            // with the entry, which is what controls the expiration
+            case EvictionReason.Capacity:
+                // Happens when the cache is full
```

Comment:
```suggestion
            // Happens when the cache is full
```

Reply:
Formatting was fighting me, so I kept it as is.